contentprovider.go

```diff
@@ -212,6 +212,47 @@ func (p *contentProvider) fillContentMatches(ms []*candidateMatch) []LineMatch {
 		}
 		finalMatch.Line = data[lineStart:lineEnd]
 
+		if numContextLines > 0 {
+			newLinesIndexes := p.newlines()
+
+			// Get context lines before the current line.
+			var before [][]byte
+			curLine := num - numContextLines - 1
+			if curLine < 0 {
+				curLine = 0
+			}
+			var startIndex uint32
+			var endIndex uint32
+			for curLine < num-1 {
+				endIndex = newLinesIndexes[curLine]
+				startIndex = 0
+				if curLine > 0 {
+					startIndex = newLinesIndexes[curLine-1]
+					before = append(before, data[startIndex+1:endIndex])
+				} else {
+					before = append(before, data[startIndex:endIndex])
+				}
+				curLine += 1
+			}
+			finalMatch.LinesBefore = before
+
+			// Get context lines after the current line.
+			var after [][]byte
+			numLines := len(newLinesIndexes)
+			curLine = num - 1
+			for curLine < num+numContextLines-1 && curLine < numLines {
+				startIndex = newLinesIndexes[curLine]
+				endIndex = p.fileSize
+				if curLine < numLines-1 {
+					endIndex = newLinesIndexes[curLine+1]
+				}
+				after = append(after, data[startIndex+1:endIndex])
+				curLine += 1
+			}
+
+			finalMatch.LinesAfter = after
+		}
```

Comment:
This code is quite verbose, I think it would benefit from a helper function maybe. Here is something I just sketched up without testing. Would allow you to share implementation between calculating before and after.

```go
func lines(data []byte, newlines []uint32, low, high int) [][]byte {
    if len(newlines) == 0 {
        return nil
    }

    if low < 0 {
        low = 0
    }
    if high > len(newlines) {
        high = len(newlines)
    }

    var lines [][]byte
    for i := low; i < high; i++ {
        begin := newlines[i] + 1
        if begin > len(data) {
            break
        }

        var end uint32
        if i + 1 < len(newlines) {
            end = newlines[i+1]
        } else {
            end = uint32(len(data))
        }

        lines = append(lines, data[begin:end])
    }

    return lines
}